File formats
=============================

Input
-------------------------

HiC map
............

a bin-contact pair format
^^^^^^^^^^^

-in\_bin
''''''''

A bin file defines the chromosome, start and end positions of each bin, an example: `fly_30k.cbins <https://github.com/changlabtw/hicmaptools/blob/master/examples/fly_30k.cbins>`_ (D.mel in 30k bin). 

::

 cbin   chr from.coord  to.coord
  1     2L  6000        7000
  2     2L  7000        8000     
  3     2L  8000        9000     
  4     2L  9000        10000     
  5     2L  12000       13000


-in\_map 
''''''''

A contact map contains Hi-C intensity indexed by bins in text format, an example `fly_30k.n_contact <https://github.com/changlabtw/hicmaptools/blob/master/examples/fly_30k.n_contact>`_.

::

 cbin1   cbin2   expected_count  observed_count     
  1   1   0.077080    50     
  1   2   0.389912    314     
  1   3   0.493750    163     
  1   4   0.560505    169     
  1   5   0.368884    79

- expected\_count : the expected contact between those two genome locis (bins) according to model. It could be 1 if no model is applied.
- observed\_count : the observed contact between those two genome locis (bins) in Hi-C data

It takes time to parse text format so we develop ``bin/genBinMap`` to turn text format (.n_contact) into binary format (.binmap) which could be used in -in\_map, an example `fly_30k.binmap <https://github.com/changlabtw/hicmaptools/blob/master/examples/fly_30k.binmap>`_.

::
 
 genBinMap [options] -in_ncontact input.n_contact -out_binmap out.binmap
 >bin/genBinMap -in_ncontact examples/fly_30k.n_contact -out_binmap examples/fly_30k.binmap
 >hicmaptools -in_map examples/fly_30k.binmap -in_bin examples/fly_30k.cbins -bat examples/bat.bed -output batTest.tsv 

.hic format
^^^^^^^^^^^
The . *hic* format is generated by `Juicer <https://github.com/aidenlab/juicer>`_. The parser of . *hic* is adapted from `straw <https://github.com/aidenlab/straw>`_.

-in\_hic
''''''''

A .hic file

-in\_hic\_norm (optional)
''''''''

To extract data based on a specified normalization method, NONE, VC, VC_SQRT, or KR (default: NONE).

-in\_hic\_resol (optional)
''''''''

To extract data at a specified resolution, e.g., 5000, 10000, or 50000 (default: 10000).


Query file
............

The query file is in `bed format <https://genome.ucsc.edu/FAQ/FAQformat.html#format1>`_ where the first three columns are enough.
The example file of each query mode is listed:

- bat: `bat.bed <https://github.com/changlabtw/hicmaptools/blob/master/examples/bat.bed>`_ a PRE binding site
- local: `local.bed <https://github.com/changlabtw/hicmaptools/blob/master/examples/local.bed>`_ a PcG TAD
- loop: `loop.bed <https://github.com/changlabtw/hicmaptools/blob/master/examples/loop.bed>`_ gene, Antp
- pair: `pair.bed <https://github.com/changlabtw/hicmaptools/blob/master/examples/pair.bed>`_ a pair of insulator binding sites
- sites: `sites.bed <https://github.com/changlabtw/hicmaptools/blob/master/examples/sites.bed>`_ a list of insulator binding sites in range 3L:10000000-11000000
- submap: `submap.bed <https://github.com/changlabtw/hicmaptools/blob/master/examples/submap.bed>`_ a region contains Antp-BX long range conttact
- TAD: `TAD.bed <https://github.com/changlabtw/hicmaptools/blob/master/examples/bat.bed>`_ selected TADs in range 2L:2000000~3000000

----------------------------

Output
-------------------------

There are two output files. You can use the tool ``tools/visualizeRandomTest.R`` to examine query's contact intensity aganist the null hypothesis (`Shuffle test <random.html>`_).

- output.tsv : the contact intensity of you interested where 
 - sum_* indicates intensity of HiC
 - rand_* indicates intensity of shuffle test
 - divide_* indicates ratio of sum/rand
 - rank_* indicates the rank of HiC among shuffle test

::

 index	chrom	start	end	sum_obs	sum_exp	sum_nor	rand_obs	rand_exp	rand_nor	divide_obs	divide_exp	divide_nor	rank_obs	rank_exp	rank_nor	
 1	2L	594629	595145	47916.000	459.715	2380.531	32618.180	314.679	2525.479	1.469	1.461	0.943	0.100	0.140	0.600

- output **_random** .txt : the observed, expected and normalizated contact intensities of the null hypothesis starting from the third row where the second row is the query intensity

::

 random_obs,random_exp,random_nor
 47916,459.715,2380.53
 19632,158.539,2956.25
 57574,448.25,2832.44
 7074,60.7897,3029.22
 33009,246.588,3311.8

